File: 0078_Subsets/0061_Rotate List/Solution.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    @staticmethod
    def deserialize(data):
        """Convert a string list representation to a linked list."""
        if not data or data == "[]":
            return None
        values = list(map(int, data.strip("[]").split(",")))
        dummy = ListNode(0)
        current = dummy
        for val in values:
            current.next = ListNode(val)
            current = current.next
        return dummy.next

    @staticmethod
    def serialize(head):
        """Convert a linked list to a string list representation."""
        if not head:
            return "[]"
        result = []
        current = head
        while current:
            result.append(str(current.val))
            current = current.next
        return "[" + ",".join(result) + "]"


class Solution:
    def rotateRight(self, head, k):
        if not head or not head.next or k == 0:
            return head

        # Calculate the length of the linked list
        length = 1
        current = head
        while current.next:
            current = current.next
            length += 1

        # Connect the tail to the head to form a circular list
        current.next = head

        # Find the new head after rotation
        k = k % length  # Normalize k to avoid unnecessary rotations
        steps_to_new_head = length - k
        new_tail = head
        for _ in range(steps_to_new_head - 1):
            new_tail = new_tail.next

        new_head = new_tail.next
        new_tail.next = None  # Break the circular list

        return new_head


# Example Usage
if __name__ == "__main__":
    input_list = "[1,2,3,4,5]"
    k = 2

    head = ListNode.deserialize(input_list)
    solution = Solution()
    rotated_head = solution.rotateRight(head, k)
    output = ListNode.serialize(rotated_head)

    print(output)  # Output: [4,5,1,2,3]

File: 0078_Subsets/0061_Rotate List/Solution.java
class ListNode {
    int val;
    ListNode next;

    ListNode() {}
    ListNode(int val) {
        this.val = val;
    }
    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

    // Deserialize: Convert a string to a linked list
    public static ListNode deserialize(String data) {
        if (data == null || data.equals("[]")) {
            return null;
        }

        data = data.substring(1, data.length() - 1); // Remove brackets
        String[] values = data.split(",");
        ListNode dummy = new ListNode(0);
        ListNode current = dummy;

        for (String val : values) {
            current.next = new ListNode(Integer.parseInt(val.trim()));
            current = current.next;
        }

        return dummy.next;
    }

    // Serialize: Convert a linked list to a string
    public static String serialize(ListNode head) {
        if (head == null) {
            return "[]";
        }

        StringBuilder sb = new StringBuilder();
        sb.append("[");

        ListNode current = head;
        while (current != null) {
            sb.append(current.val);
            if (current.next != null) {
                sb.append(",");
            }
            current = current.next;
        }

        sb.append("]");
        return sb.toString();
    }
}

public class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) {
            return head;
        }

        // Calculate the length of the list
        ListNode current = head;
        int length = 1;
        while (current.next != null) {
            current = current.next;
            length++;
        }

        // Connect the tail to the head to form a circular list
        current.next = head;

        // Find the new head
        k = k % length; // Normalize k to avoid unnecessary rotations
        int stepsToNewHead = length - k;
        ListNode newTail = head;
        for (int i = 1; i < stepsToNewHead; i++) {
            newTail = newTail.next;
        }

        // Break the circular list
        ListNode newHead = newTail.next;
        newTail.next = null;

        return newHead;
    }

    public static void main(String[] args) {
        // Example Input
        String input = "[1,2,3,4,5]";
        int k = 2;

        // Deserialize input
        ListNode head = ListNode.deserialize(input);

        // Perform rotation
        Solution solution = new Solution();
        ListNode rotatedHead = solution.rotateRight(head, k);

        // Serialize and print output
        String output = ListNode.serialize(rotatedHead);
        System.out.println(output); // Output: [4,5,1,2,3]
    }
}
